src/management-system-v2/app/(dashboard)/[environmentId]/executions/[processId]/instance-coloring.ts

```diff
@@ -0,0 +1,160 @@
+import { BPMNCanvasRef } from '@/components/bpmn-canvas';
+import { InstanceInfo } from '@/lib/engines/deployment';
+import { getMetaDataFromElement } from '@proceed/bpmn-helper/src/getters';
+import { getPlannedEnd } from './instance-helpers';
+
+export const colorOptions = [
+  {
+    key: 'processColors',
+    label: 'Process Colors',
+  },
+  {
+    key: 'noColors',
+    label: 'No Colors',
+  },
+  {
+    key: 'timeColors',
+    label: 'Time Colors',
+  },
+  {
+    key: 'executionColors',
+    label: 'Execution Colors',
+  },
+] as const;
+export type ColorOptions = (typeof colorOptions)[number]['key'];
+
+// NOTE: this would break if some of these functions are used somewhere else
+let appliedStyling: { elementId: string; color: string }[] = [];
+
+export function flushPreviousStyling() {
+  appliedStyling = [];
+}
+
+export function applyColors(
+  bpmnViewer: BPMNCanvasRef,
+  instance: InstanceInfo,
+  colors: ColorOptions,
+) {
+  const canvas = bpmnViewer.getCanvas();
+
+  // remove previous styling
+  for (const { elementId, color } of appliedStyling) canvas.removeMarker(elementId, color);
+
+  // apply new styling
+  appliedStyling = flowElementsStyling(bpmnViewer, instance, colors);
+  for (const { elementId, color } of appliedStyling) canvas.addMarker(elementId, color);
+}
+
+function getExecutionColor(executionState: string, wasInterrupted: boolean) {
+  switch (executionState) {
+    case 'COMPLETED':
+      return wasInterrupted ? 'yellow' : 'green';
+    case 'ERROR-SEMANTIC':
+    case 'ERROR-TECHNICAL':
+    case 'ERROR-CONSTRAINT_UNFULFILLED':
+    case 'STOPPED':
+    case 'ABORTED':
+    case 'FAILED':
+    case 'TERMINATED':
+      return 'red';
+    default:
+      // SKIPPED
+      return 'white';
+  }
+}
+
+export function progressToColor(
+  elementMetaData: any,
+  timeInfo: { startTime: number | null; endTime: number | null },
+) {
+  let plannedStart = timeInfo.startTime;
+  if (elementMetaData.timePlannedOccurrence) {
+    plannedStart = new Date(elementMetaData.timePlannedOccurrence).getTime();
+  }
+
+  const plannedDuration = elementMetaData.timePlannedDuration;
+
+  if (!plannedStart || !plannedDuration) return 'white';
+
+  const plannedEnd = getPlannedEnd(plannedStart, plannedDuration);
+  if (!plannedEnd) return 'white';
+
+  if (timeInfo.endTime) {
+    if (plannedEnd.getTime() > timeInfo.endTime) return 'green';
+    else return 'red';
+  }
+
+  const durationInMs = plannedEnd.getTime() - plannedStart;
+  const criticalTime = Math.floor(0.7 * durationInMs);
+  const currentDate = new Date().getTime();
+
+  if (currentDate < plannedStart + criticalTime) {
+    return 'green';
+  } else if (currentDate < plannedEnd.getTime()) {
+    return 'orange';
+  } else {
+    return 'red';
+  }
+}
+
+function flowElementsStyling(
+  bpmnViewer: BPMNCanvasRef,
+  instance: InstanceInfo,
+  colors: ColorOptions,
+) {
+  const instanceFlowElements = bpmnViewer.getAllElements();
```

Comment:
There are problems with the colors (which partially also exist in the old MS).
- If you select "No Colors" it will only remove the fill color but not the stroke color.
- Time colors seems not to work correctly. If you give a task a duration and let the task run for longer than that duration the task should be colored red
- Execution Colors keeps the stroke color of executed elements and the fill color of not yet executed elements